source/git/source.go

```diff
@@ -635,6 +654,16 @@ func (gs *gitSourceHandler) Snapshot(ctx context.Context, g session.Group) (out
 	}
 
 	git = git.New(gitutil.WithWorkTree(checkoutDir), gitutil.WithGitDir(gitDir))
+	if gs.src.Checksum != "" {
```

Comment:
Same as https://github.com/moby/buildkit/pull/5903#discussion_r2067877210 . Why are we doing this after the checkout/fetch instead of doing it before?

Reply:
Because post-checkout verification is easier to cover multiple if-else conditions

Reply:
I don't think this is correct actually. Why is this left the `HEAD` when `KeepGitDir`? Should always be `ref`. You are running this in the context of original `git` repo, not the checkout (what may lack git dir at all).

It still makes most sense to me to check that the checksum of the ref matches in the origin repo before trying to do the fetch or checkout and stopping if it doesn't. I don't know what `if-else` you mean as this should be done before the `keep-git-dir` check and doesn't need to be done separately based on the `keep-git-dir` value.

Reply:
Ok, I understand why this works. There is `gitDir = checkoutDirGit` in line 613. But I moved it and still seems to work fine. Also added more test coverage. PTAL https://github.com/moby/buildkit/compare/master...tonistiigi:buildkit:git-checksum2?expand=1

Reply:
Thanks, cherry-picked your commit